metrics_utility/automation_controller_billing/collectors.py

```diff
@@ -563,42 +571,89 @@ def total_workers_vcpu(since, full_path, until, **kwargs):
         # This message must always appear in the log regardless of the log level.
         logger_info_level.info(json.dumps(info, indent=2))
-        return {'cluster_name': info['cluster_name'], 'total_workers_vcpu': info['total_workers_vcpu']}
+        return {'timestamp': info['timestamp'], 'cluster_name': info['cluster_name'], 'total_workers_vcpu': info['total_workers_vcpu']}
+
+    url = os.getenv('METRICS_UTILITY_PROMETHEUS_URL')
+    if not url:
+        prometheus_default_url = 'https://prometheus-k8s.openshift-monitoring.svc.cluster.local:9091'
+        logger.info(
+            f'environment variable METRICS_UTILITY_PROMETHEUS_URL is not set, \
+                    default {prometheus_default_url} will be assigned'
+        )
+        url = prometheus_default_url
 
     try:
-        kube_config.load_incluster_config()
-    except kube_config.ConfigException:
-        try:
-            kube_config.load_kube_config()
-        except kube_config.ConfigException as e:
-            logger.error(f'Could not configure Kubernetes Python client ERROR: {e}')
-            raise MetricsException(f'Could not configure Kubernetes Python client ERROR: {e}')
-
-    # Create a CoreV1Api client
-    api_instance = client.CoreV1Api()
-    if not api_instance:
-        raise MetricsException('Could not get a Kube CoreV1Api client')
+        prom = PrometheusClient(url=url)
+    except Exception as e:
+        raise MetricsException(f'Can not create a prometheus api client ERROR: {e}')
 
     try:
-        nodes = api_instance.list_node()
-    except Exception as e:
+        total_workers_vcpu, promql_query = get_total_workers_cpu(prom, prev_hour_start)
+        timeline = get_cpu_timeline(prom, prev_hour_start, prev_hour_end)
+    except MetricsException as e:
         raise MetricsException(f'Unexpected error when retrieving nodes: {e}')
 
-    if nodes is None:
-        raise MetricsException('No nodes found')
+    info['promql_query'] = promql_query
+    info['timeline'] = timeline
+
+    logger.debug(f'total_workers_vcpu: {total_workers_vcpu}')
 
-    total_workers_vcpu = 0
-    # In SaaS case we have only Worker nodes and so we don't need to filter out the control plan.
-    # If it used for other environement, we might need to implement the filtering.
-    for node_info in nodes.items:
-        for resource, value in node_info.status.capacity.items():
-            if resource == 'cpu':
-                info['nodes'].append({node_info.metadata.name: int(value)})
-                total_workers_vcpu += int(value)
+    # This can happen when the prev_hour_start doesn't have data, it could be when the cluster just started or
+    # if for some reasons prometheus loss some data.
+    if total_workers_vcpu is None:
+        logger.warning('No data availble yet, the cluster is probably running for less than an hour')
+        raise MetricsException('No data availble yet, the cluster is probably running for less than an hour')
```

Comment:
Nice catch. This seems sub-optimal given we'll hit this for any instance that is freshly provisioned vs assigned from the pool. What's in the `timeline` in this case? Is it also empty? I hate for us to code an error for a fairly normal circumstance. What could we do to improve this without making the implementation overly complex? This could be opened as a follow-up issue if need be.

Reply:
The timeline start at instant t and as there is no data at instant t, I guess we will get no timeline at all. but I will confirm, no easy to test.

Reply:
Let's open a follow-up issue to better support the first partial hour and discuss with the team the relative priority of fixing this. It will depend on how often we're using non-pooled instances and how we're monitoring for metering issues in production. But we should be able to determine when metrics first appeared and use that as the lower bound for the first collection.